# Sales Data Analysis Using NumPy

## Project Overview

This project analyzes a sales dataset using NumPy to strengthen fundamental numerical computing and data analysis skills.

The dataset contains information about sales transactions, including sales amount, profit, quantity, product categories, payment methods, customers, and geographic information.

The analysis will focus on understanding the numerical structure of the dataset and using NumPy operations to calculate important sales metrics.

## Objectives

The main objectives of this project are:

- Understand NumPy arrays
- Inspect numerical sales data
- Calculate total sales
- Calculate total profit
- Calculate average sales
- Analyze quantities sold
- Identify minimum and maximum values
- Use NumPy indexing and slicing
- Apply Boolean filtering
- Compare sales and profit
- Analyze numerical relationships
- Extract business insights from the data

## Tools

- Python
- NumPy
- Pandas
- Matplotlib

## Dataset

The dataset contains sales transaction information such as:

- Order ID
- Sales Amount
- Profit
- Quantity
- Category
- Sub-Category
- Payment Mode
- Order Date
- Customer
- State
- City

The dataset is used for learning and exploratory analysis.

## 1. Import Required Libraries

NumPy is the primary library used in this project for numerical analysis.

Pandas is used initially to load the CSV dataset because the dataset contains both numerical and categorical columns.

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## 2. Load the Dataset

The sales dataset is stored as a CSV file.

We first load the dataset using Pandas so that we can inspect its structure and then extract numerical columns for NumPy analysis.

In [3]:
af = pd.read_csv("Sales Dataset.csv")

In [ ]:
af.shape

(1194, 12)

In [6]:
af.head()

,Order ID,Amount,Profit,Quantity,Category,Sub-Category,PaymentMode,Order Date,CustomerName,State,City,Year-Month
0,B-26776,9726,1275,5,Electronics,Electronic Games,UPI,2023-06-27,David Padilla,Florida,Miami,2023-06
1,B-26776,9726,1275,5,Electronics,Electronic Games,UPI,2024-12-27,Connor Morgan,Illinois,Chicago,2024-12
2,B-26776,9726,1275,5,Electronics,Electronic Games,UPI,2021-07-25,Robert Stone,New York,Buffalo,2021-07
3,B-26776,4975,1330,14,Electronics,Printers,UPI,2023-06-27,David Padilla,Florida,Miami,2023-06
4,B-26776,4975,1330,14,Electronics,Printers,UPI,2024-12-27,Connor Morgan,Illinois,Chicago,2024-12


In [7]:
af.columns

Index(['Order ID', 'Amount', 'Profit', 'Quantity', 'Category', 'Sub-Category',
       'PaymentMode', 'Order Date', 'CustomerName', 'State', 'City',
       'Year-Month'],
      dtype='str')

In [8]:
af.info()

<class 'pandas.DataFrame'>
RangeIndex: 1194 entries, 0 to 1193
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Order ID      1194 non-null   str  
 1   Amount        1194 non-null   int64
 2   Profit        1194 non-null   int64
 3   Quantity      1194 non-null   int64
 4   Category      1194 non-null   str  
 5   Sub-Category  1194 non-null   str  
 6   PaymentMode   1194 non-null   str  
 7   Order Date    1194 non-null   str  
 8   CustomerName  1194 non-null   str  
 9   State         1194 non-null   str  
 10  City          1194 non-null   str  
 11  Year-Month    1194 non-null   str  
dtypes: int64(3), str(9)
memory usage: 112.1 KB


## 3. Identify Numerical Variables

For the NumPy analysis, we will focus on the numerical variables in the dataset.

The main numerical variables are:

- Amount
- Profit
- Quantity

These variables allow us to perform numerical calculations using NumPy arrays.

In [9]:
af[["Amount", "Profit", "Quantity"]].head()

,Amount,Profit,Quantity
0,9726,1275,5
1,9726,1275,5
2,9726,1275,5
3,4975,1330,14
4,4975,1330,14


In [11]:
amount = af["Amount"].to_numpy()
profit = af["Profit"].to_numpy()
quantity = af["Quantity"].to_numpy()

print("Amount:", amount[:10])
print("Profit:", profit[:10])
print("Quantity:", quantity[:10])

Amount: [9726 9726 9726 4975 4975 4975 1525 1525  883 8127]
Profit: [1275 1275 1275 1330 1330 1330  185  185  117 3551]
Quantity: [ 5  5  5 14 14 14 12 12 10 16]


## 4. Understanding NumPy Arrays

After extracting the numerical columns from the DataFrame, we can examine the structure of the resulting NumPy arrays.

We will inspect:

- Number of observations
- Number of dimensions
- Shape
- Data type

In [12]:
print("Amount")
print("Size:", amount.size)
print("Dimensions:", amount.ndim)
print("Shape:", amount.shape)
print("Data type:", amount.dtype)

Amount
Size: 1194
Dimensions: 1
Shape: (1194,)
Data type: int64


In [13]:
print("\nProfit")
print("Size:", profit.size)
print("Dimensions:", profit.ndim)
print("Shape:", profit.shape)
print("Data type:", profit.dtype)

print("\nQuantity")
print("Size:", quantity.size)
print("Dimensions:", quantity.ndim)
print("Shape:", quantity.shape)
print("Data type:", quantity.dtype)


Profit
Size: 1194
Dimensions: 1
Shape: (1194,)
Data type: int64

Quantity
Size: 1194
Dimensions: 1
Shape: (1194,)
Data type: int64


## 5. Total Sales

The total sales amount represents the combined sales value across all transactions.

NumPy's `sum()` function can be used to calculate the total sales.

In [16]:
total_sales = np.sum(amount)

print("Total Profit" , total_sales)

Total Profit 6182639


In [15]:
total_profit = np.sum(profit)

print("Total Profit" , total_profit)

Total Profit 1610697


In [17]:
total_quantity = np.sum(quantity)

print("Total Quantity" , total_quantity)

Total Quantity 12745


In [18]:
average_sales = np.mean(amount)

print("Average Sales" ,average_sales)

Average Sales 5178.089614740368


In [19]:
average_profit = np.mean(profit)

print("Average Profit" , average_profit)

Average Profit 1348.9924623115578


In [24]:
average_quantity = np.mean(quantity)

print("Average Quantity", average_quantity)

Average Quantity 10.674204355108877


In [25]:
print("Total Sales:", total_sales)
print("Total Profit:", total_profit)
print("Total Quantity:", total_quantity)

print("Average Sales:", average_sales)
print("Average Profit:", average_profit)
print("Average Quantity:", average_quantity)

Total Sales: 6182639
Total Profit: 1610697
Total Quantity: 12745
Average Sales: 5178.089614740368
Average Profit: 1348.9924623115578
Average Quantity: 10.674204355108877


## Insight

The dataset contains a total sales value of 6,182,639 and a total profit of 1,610,697 across 12,745 units sold.

On average, each transaction generated approximately:

- Sales Amount: 5,178.09
- Profit: 1,348.99
- Quantity Sold: 10.67 units

These metrics provide an initial understanding of the business performance and establish a baseline for further analysis of sales and profitability.

In [26]:
min_sales = np.min(amount)
max_sales = np.max(amount)

print("Minimum Sales:", min_sales)
print("Maximum Sales:", max_sales)

Minimum Sales: 508
Maximum Sales: 9992


In [27]:
min_profit = np.min(profit)
max_profit = np.max(profit)

print("Minimum Profit:", min_profit)
print("Maximum Profit:", max_profit)

Minimum Profit: 50
Maximum Profit: 4930


In [28]:
min_quantity = np.min(quantity)
max_quantity = np.max(quantity)

print("Minimum Quantity:", min_quantity)
print("Maximum Quantity:", max_quantity)

Minimum Quantity: 1
Maximum Quantity: 20


In [29]:
sales_max_index = np.argmax(amount)
sales_min_index = np.argmin(amount)

print("Highest Sales Position:", sales_max_index)
print("Lowest Sales Position:", sales_min_index)

Highest Sales Position: 120
Lowest Sales Position: 749


In [30]:
af.iloc[sales_max_index]

Order ID                 B-25936
Amount                      9992
Profit                      3696
Quantity                      20
Category         Office Supplies
Sub-Category                Pens
PaymentMode           Debit Card
Order Date            2021-10-09
CustomerName    Patrick Williams
State                   New York
City                   Rochester
Year-Month               2021-10
Name: 120, dtype: object

In [31]:
np.min(amount)
np.max(amount)

np.min(profit)
np.max(profit)

np.min(quantity)
np.max(quantity)

np.int64(20)

In [32]:
sales_max_index = np.argmax(amount)
sales_min_index = np.argmin(amount)

af.iloc[sales_max_index]
af.iloc[sales_min_index]

Order ID           B-25465
Amount                 508
Profit                 177
Quantity                17
Category         Furniture
Sub-Category         Sofas
PaymentMode            UPI
Order Date      2022-03-23
CustomerName      Tony Ray
State                 Ohio
City             Cleveland
Year-Month         2022-03
Name: 749, dtype: object

### Insight

The minimum sales transaction generated sales of 508, while the corresponding transaction generated a profit of 177.

The transaction involved 17 units of a Furniture product from the Sofas sub-category.

This demonstrates how `np.argmin()` can be used to identify the position of the minimum value in a NumPy array, after which the corresponding record can be retrieved from the original DataFrame using `.iloc[]`.

# Conclusion

This project provided a practical introduction to using NumPy for numerical sales analysis.

Using the sales dataset, I converted important numerical columns such as Amount, Profit, and Quantity into NumPy arrays and applied fundamental NumPy operations to analyze the data.

The analysis showed:

- Total Sales: **6,182,639**
- Total Profit: **1,610,697**
- Total Quantity Sold: **12,745 units**
- Average Sales per Transaction: **5,178.09**
- Average Profit per Transaction: **1,348.99**
- Average Quantity per Transaction: **10.67 units**

NumPy functions such as `np.sum()`, `np.mean()`, `np.min()`, `np.max()`, `np.argmax()`, and `np.argmin()` were used to calculate important business metrics and identify extreme transactions.

The project also demonstrated how NumPy array positions can be connected back to the original Pandas DataFrame using `.iloc[]`. This is an important concept when combining numerical analysis with real-world datasets.

Overall, this project strengthened my understanding of NumPy arrays, numerical calculations, indexing, and basic data analysis while applying these concepts to a real sales dataset.

# Key Insights

1. **Strong overall sales volume**

   The dataset contains total sales of **6.18 million**, indicating substantial transaction activity across the recorded period.

2. **Positive overall profitability**

   Total profit is approximately **1.61 million**, showing that the transactions in the dataset generated positive overall profit.

3. **Average transaction performance**

   An average transaction generated approximately **5,178 in sales** and **1,349 in profit**, providing a useful baseline for evaluating individual transactions.

4. **Quantity sold**

   A total of **12,745 units** were sold, with an average of approximately **10.67 units per transaction**.

5. **Extreme transactions can be identified with NumPy**

   `np.argmax()` and `np.argmin()` made it possible to locate the positions of the highest and lowest sales transactions. These positions were then used with Pandas `.iloc[]` to inspect the complete transaction records.

6. **NumPy is useful for numerical analysis**

   The project demonstrated that NumPy provides efficient and simple tools for performing calculations on numerical data, making it an important foundation for data analysis and data science.

7. **Foundation for further analysis**

   The numerical analysis performed in this project can be extended into more advanced analysis, such as category performance, regional sales, monthly trends, profit margins, and visualization.